js/src/wrappers/claude-agent-sdk.ts

```diff
@@ -0,0 +1,505 @@
+  description: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  inputSchema: any;
+  handler: ToolHandler<T>;
+};
+
+/**
+ * Wraps the Claude Agent SDK's query function to add Braintrust tracing.
+ * Traces the entire agent interaction including all streaming messages.
+ *
+ * @param queryFn - The original query function from @anthropic-ai/claude-agent-sdk
+ * @returns A wrapped query function with tracing
+ *
+ * @example
+ * ```typescript
+ * import { query } from "@anthropic-ai/claude-agent-sdk";
+ * import { wrapClaudeAgentQuery } from "braintrust";
+ *
+ * const wrappedQuery = wrapClaudeAgentQuery(query);
+ *
+ * const result = wrappedQuery({
+ *   prompt: "Write a hello world function",
+ *   options: { model: "claude-3-5-sonnet-20241022" }
+ * });
+ *
+ * for await (const message of result) {
+ *   console.log(message);
+ * }
+ * ```
+ */
+export function wrapClaudeAgentQuery<
+  T extends (...args: unknown[]) => AsyncGenerator<SDKMessage, void, unknown>,
+>(queryFn: T, defaultThis?: unknown): T {
+  const proxy: T = new Proxy(queryFn, {
+    apply(target, thisArg, argArray) {
+      const params = (argArray[0] ?? {}) as {
+        prompt?: string | AsyncIterable<SDKMessage>;
+        options?: QueryOptions;
+      };
+
+      const { prompt, options = {} } = params;
+
+      const span = startSpan({
+        name: "Claude Agent",
+        spanAttributes: {
+          type: SpanTypeAttribute.TASK,
+        },
+        event: {
+          input:
+            typeof prompt === "string"
+              ? prompt
+              : { type: "streaming", description: "AsyncIterable<SDKMessage>" },
+          metadata: options,
+        },
+      });
+
+      let ended = false;
+      const finalResults: Array<{ content: unknown; role: string }> = [];
+
+      // Track messages by message.message.id to group streaming updates
+      let currentMessageId: string | undefined;
+      let currentMessageStartTime = getCurrentUnixTimestamp();
+      const currentMessages: SDKMessage[] = [];
+
+      const endSpan = () => {
+        if (!ended) {
+          ended = true;
+          span.end();
+        }
+      };
+
+      // Create an LLM span for accumulated messages with the same message ID.
+      // LLM spans can contain multiple streaming message updates. We create the span
+      // when we proceed to a new message ID or when the query completes.
+      const createLLMSpan = async () => {
+        const finalMessageContent = await _createLLMSpanForMessages(
+          currentMessages,
+          prompt,
+          finalResults,
+          options,
+          currentMessageStartTime,
+          await span.export(),
+        );
+
+        if (finalMessageContent) {
+          finalResults.push(finalMessageContent);
+        }
+
+        currentMessages.length = 0;
+      };
+
+      // Create wrapped async generator that maintains span context
+      const wrappedGenerator: AsyncGenerator<SDKMessage, void, unknown> =
```

Comment:
When I'm creating new integrations I always try to reuse the `start_span` context manager (in Python) and in JS land use the `traced` method. Did you happen to try `traced` here? I'm curious how it would handle `AsyncGenerator`.

Reply:
Oh there's a `wrapTracedAsyncGenerator`

Reply:
I'm guessing there's just a lot to handle here compared to the default behavior of the traced methods